### PR TITLE
Show inserter on slash: /+SEARCH

### PR DESCRIPTION
**Pull request by ellatrix**

### CODE REVIEW

**Comment by iseulde** [comment]
Something is still going wrong with the CSS here. Maybe because I'm cross-importing and it causes this?

I wonder how I should improve this? Move `InserterMenu` to components? But then it still needs to be state aware? Make a special blocks folder in `editor/` that is state aware (internal)? Note that this is not the only block, there will be blocks like table of contents that will also need to be state aware. Cc @youknowriad @aduth.

**Comment by mtias** [comment]
@iseulde I think we should do a separate inserter component for this case—that just lists blocks in a single column list—instead of reusing the other.

**Comment by mtias** [comment]
Probably call it blocks-autcomplete. And it would look closer to:

image

**Comment by mtias** [comment]
We should use the same design for eventually autocompleting tags, users, etc.

**Comment by iseulde** [comment]
@mtias Ok, I'll adjust

**Comment by jasmussen** [comment]
Love this, thank you for working on it. 

I would agree with Matías, for this round 2 of the slash command, let's try adhering as closely as possible to the pattern from whence the inspiration came: Slack and IRC:

slashcommand

File: blocks/editable/patterns.js [hunk 1]
```diff
@@ -219,6 +221,37 @@ export default function( editor ) {
 		onReplace( [ block ] );
 	}
 
+	function character() {
+		if ( ! onReplace ) {
+			return;
+		}
+
+		// Merge text nodes.
+		editor.getBody().normalize();
```
Comment (aduth) [inline blocks/editable/patterns.js]: How expensive an operation is this for something which will occur on every single keypress? _Edit:_ I guess it's every keypress for a block which implements `onReplace` (inserter block).

File: editor/inserter/menu.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { __, _n, sprintf } from '@wordpress/i18n';
 import { Component } from '@wordpress/element';
 import { Popover, withFocusReturn, withInstanceId, withSpokenMessages } from '@wordpress/components';
 import { keycodes } from '@wordpress/utils';
-import { getCategories, getBlockTypes, BlockIcon } from '@wordpress/blocks';
+import { getCategories, getBlockTypes, BlockIcon } from '../../blocks';
```
Comment (aduth) [inline editor/inserter/menu.js]: Why this change?

File: blocks/library/inserter/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * Internal dependencies
+ */
+import './block.scss';
+import { registerBlockType, createBlock } from '../../api';
+import InserterMenu from '../../../editor/inserter/menu';
+
+const { BACKSPACE, ESCAPE } = keycodes;
+
+registerBlockType( 'core/inserter', {
```
Comment (aduth) [inline blocks/library/inserter/index.js]: I've not tested this, but what's the expected behavior with this being implemented as a block? A user has to insert the inserter block, then they have the option to `/` to trigger the menu? I'd not really expected this to be a block, but rather something built-in to either Editable or the default paragraph block.

**Review by aduth — COMMENTED** [review]
>I wonder how I should improve this? Move `InserterMenu` to components? But then it still needs to be state aware? Make a special blocks folder in `editor/` that is state aware (internal)? Note that this is not the only block, there will be blocks like table of contents that will also need to be state aware. Cc @youknowriad @aduth.

I agree, we will likely need better support for blocks components and perhaps blocks themselves having access to and manipulating state. Remarked previously at at https://github.com/WordPress/gutenberg/pull/2119#issuecomment-319912820 and https://github.com/WordPress/gutenberg/pull/2182#issuecomment-319973243 , I question whether our split between blocks and editor is sensible. I think there are _post-editor_ specific concerns that might not be relevant to blocks and editors generally, but the idea of a block and editor should be unified, at which point the core of an editor consists of block APIs, components, and editor state itself. It becomes the editor chrome (headers, sidebars, etc) which might need to be externalized as a layer atop this core.

**Comment by aduth** [comment]
Superseded by #2630
